go/enclave/evm/evm_facade.go

```diff
@@ -104,23 +107,63 @@ func (exec *evmExecutor) execute(tx *common.L2PricedTransaction, from gethcommon
 		return nil, err
 	}
 
+	actualGasLimit := msg.GasLimit // before l1 cost is applied
+
 	receipt, err := adjustPublishingCostGas(tx, msg, s, header, noBaseFee, func() (*types.Receipt, error) {
 		s.SetTxContext(tx.Tx.Hash(), tCount)
 		return gethcore.ApplyTransactionWithEVM(msg, gp, s, header.Number, header.Hash(), header.Time, tx.Tx, usedGas, evmEnv)
 	})
 	if err != nil {
 		return nil, err
 	}
+	if receipt == nil {
+		return nil, fmt.Errorf("internal: nil receipt returned from execution")
+	}
 	receipt.Logs = s.GetLogs(tx.Tx.Hash(), header.Number.Uint64(), header.Hash(), header.Time)
 
 	contractsWithVisibility := make(map[gethcommon.Address]*core.ContractVisibilityConfig)
+
+	// Compute leftover user gas after the main execution
+	var gasLeft uint64 = actualGasLimit
+	if receipt.GasUsed > gasLeft {
+		// ensure no bugs going overboard with the l1 publishing
+		// or execution before we pick up the leftover gas.
+		return nil, fmt.Errorf("internal: gasUsed (%d) exceeds tx gasLimit (%d)", receipt.GasUsed, gasLeft)
+	}
+	gasLeft -= receipt.GasUsed
+
 	for _, contractAddress := range createdContracts {
-		var err1 error
-		contractsWithVisibility[*contractAddress], err1 = exec.visibilityReader.ReadVisibilityConfig(context.Background(), evmEnv, *contractAddress)
+		if gasLeft == 0 {
+			return nil, fmt.Errorf("out of gas while reading visibility for %s", contractAddress.Hex())
+		}
+
+		// Cap, reserve, call, and refund using helper
+		cap := min(gasLeft, maxGasForVisibility)
+		cfg, visUsed, err1 := exec.readVisibilityWithCap(context.Background(), evmEnv, gp, *contractAddress, cap)
```

Comment:
you're doing the capping in the visibility_config as well. Let's remove from here